[libc++] Update `<source_location>` and `msvc_stdlib_force_include.h`

StephanTLavavej opened the pull request:
Thanks to @mstorsjo's observation in https://github.com/llvm/llvm-project/pull/74182#issuecomment-1837617934, here's a far less intrusive way to make libc++'s filesystem tests compatible with MSVC's STL.

In `msvc_stdlib_force_include.h` (which, as the filename indicates, is force-included by the MSVC STL test environment, but never included by the libc++ test environment), we need to define 3 more macros:

* [`_CRT_DECLARE_NONSTDC_NAMES`][msvc_docs] activates the POSIX names of `getcwd` etc. As the comment explains, we need this because we test with Clang `-fno-ms-compatibility`, which defines `__STDC__` to `1`, which causes the UCRT headers to disable the POSIX names by default.
* Then we need [`_CRT_NONSTDC_NO_WARNINGS`][msvc_docs] to avoid emitting deprecation warnings about the POSIX names.
* Finally, we need `NOMINMAX` to seal away the ancient evil.

[msvc_docs]: https://learn.microsoft.com/en-us/cpp/c-runtime-library/compatibility?view=msvc-170

Like the other macros, these are defined within `#ifndef _LIBCXX_IN_DEVCRT` because we want them only within libc++'s test suite. (Our own test suite, formerly called "devcrt", reaches into our llvm-project submodule to include this header in a few places, and we don't want these macros defined there.)

While I'm in the neighborhood, I'm also adding a "simulated" macro: `__has_feature(hwaddress_sanitizer)` as `0`.

I pushed a commit to clang-format all of `msvc_stdlib_force_include.h`. (The CI's suggested diff apparently clang-formatted just the modified lines, which led to an inconsistently gross look, so I figured it was time to format the whole thing.)

Finally, as @philnik777 requested, I'm removing `__has_builtin(__builtin_source_location)` guards, including in libc++ product code. `<version>`, `source_location.version.compile.pass.cpp` (which now passes for MSVC's STL), and `version.version.compile.pass.cpp` were regenerated, no manual changes.

StephanTLavavej committed:
Define `_CRT_DECLARE_NONSTDC_NAMES`, `_CRT_NONSTDC_NO_WARNINGS`, `NOMINMAX`.

StephanTLavavej committed:
Simulate `__has_feature(hwaddress_sanitizer)` as `0` and `__has_builtin(__builtin_source_location)` as `1`.

StephanTLavavej committed:
clang-format msvc_stdlib_force_include.h, no manual changes.

philnik777 commented on libcxx/test/support/msvc_stdlib_force_include.h:
```diff
@@ -13,98 +13,106 @@
+#  define __has_attribute(X) _MSVC_HAS_ATTRIBUTE_##X
+#  define _MSVC_HAS_ATTRIBUTE_vector_size 0
+
+#  define __has_builtin(X) _MSVC_HAS_BUILTIN_##X
```
I suspect that we don't actually need this check anymore. `__builtin_source_location` has been added in Clang 15, and I don't think we have any compiler that is based on something earlier anymore. Would you mind trying to remove it and checking whether the CI passes on that?

StephanTLavavej committed:
Attempt to remove `__has_builtin(__builtin_source_location)` guards.

<version>, source_location.version.compile.pass.cpp, and version.version.compile.pass.cpp were regenerated, no manual changes.

StephanTLavavej commented:
@philnik777 Looks like removing `__has_builtin(__builtin_source_location)` worked! I updated the PR description accordingly.
